src/main/java/de/tum/in/www1/artemis/domain/quiz/AbstractQuizSubmission.java

```diff
@@ -0,0 +1,144 @@
+        this.submittedAnswers.remove(submittedAnswer);
+        submittedAnswer.setSubmission(null);
+        return this;
+    }
+
+    /**
+     * Filters the sensitive quiz submission information for exams, if the results are not published or the user is not an instructor
+     * It sets the {@link QuizSubmission#setScoreInPoints(Double)} & {@link SubmittedAnswer#setScoreInPoints(Double)} to null for every submitted answer.
+     * Additionally it calls {@link SubmittedAnswer#filterOutCorrectAnswers()} dynamically for the correct question type.
+     *
+     * @param examResultsPublished flag indicating if the results are published, see {@link Exam#resultsPublished()}
+     * @param isAtLeastInstructor  flag indicating if the user has instructor privileges
+     */
+    public void filterForExam(boolean examResultsPublished, boolean isAtLeastInstructor) {
+        if (!(examResultsPublished || isAtLeastInstructor)) {
+            this.setScoreInPoints(null);
+            // Dynamic binding will call the right overridden method for different question types
+            this.getSubmittedAnswers().forEach(SubmittedAnswer::filterOutCorrectAnswers);
+        }
+    }
+
+    /**
+     * Get the submitted answer to the given quizQuestion
+     *
+     * @param quizQuestion the quizQuestion that the answer should belong to
+     * @return the submitted answer for this quizQuestion (null if this quizQuestion wasn't answered by this submission)
+     */
+    public SubmittedAnswer getSubmittedAnswerForQuestion(QuizQuestion quizQuestion) {
+        for (SubmittedAnswer submittedAnswer : getSubmittedAnswers()) {
+            if (quizQuestion.equals(submittedAnswer.getQuizQuestion())) {
+                return submittedAnswer;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * calculates the scores for this submission and all its submitted answers and saves them in scoreInPoints
+     *
+     * @param quizQuestions the quiz questions this submission belongs to (is needed to have values for isCorrect in answer options)
+     */
+    public void calculateAndUpdateScores(List<QuizQuestion> quizQuestions) {
+        // set scores for all questions
+        for (QuizQuestion quizQuestion : quizQuestions) {
+            // search for submitted answer for this quizQuestion
+            SubmittedAnswer submittedAnswer = getSubmittedAnswerForQuestion(quizQuestion);
+            if (submittedAnswer != null) {
+                submittedAnswer.setScoreInPoints(quizQuestion.scoreForAnswer(submittedAnswer));
+            }
+        }
+        // set total score
+        setScoreInPoints(getScoreInPoints(quizQuestions));
+    }
+
+    /**
+     * Get the score for this submission as the number of points
+     *
+     * @param quizQuestions the quiz questions of the submission
+     * @return the resulting score
+     */
+    public Double getScoreInPoints(List<QuizQuestion> quizQuestions) {
```

Comment:
There's some duplication between this and . You should be able to just call `AbstractQuizSubmission#getScoreInPoints()` from `QuizExercise#getScoreInPointsForSubmission()` right?